### PR TITLE
[READY] Lighting list cleanup.

### PR DESCRIPTION
**Pull request by MrsTonedOne**
Removed a bunch of back reference lists that were either entirely unused, or contained references that could be found within the datums contained in other lists.

Lighting corner datums now get deleted when unused.

Light sources no longer track lighting corners where the applied light rounds to 0.

Lighting corners are no longer tracked by a constant 4 item list. This was better as 4 datum vars.

~~todo: test. This is a branch from November, iirc it was testing fine, gonna do that now that i've rebased.~~ tested fine

edit:

Lighting corners are now lazy inited and deleted. they should always (and only exist) if there is a light source shining on it within range (even if the turf has no dynamic lighting). This is needed to support turfs that become lighting enabled mid round. On the plus side, they will no longer be generated on full dark turfs.

### CODE REVIEW

**Comment by MrStonedOne** [comment]
Tested, spaming fix/break lights produces no runtimes or observable bugs, spaming shuttle moves in light interesting areas produces no runtimes or observable bugs.

Test merging on terry since it could use the extra memory.

File: code/modules/lighting/lighting_source.dm [hunk 1]
```diff
@@ -103,38 +102,39 @@
...
+	. = LUM_FALLOFF(C, pixel_turf);              \
+	. *= light_power;                            \
+	var/OLD = effect_str[C];                     \
+	if (. || OLD) {                              \
+		if(. > 0) {                              \
+			effect_str[C] = .;                   \
+		} else {                                 \
+			effect_str -= C                      \
+		}                                        \
+    	                                         \
```
Comment (MrStonedOne) [inline code/modules/lighting/lighting_source.dm:115]:
```suggestion
		                                           \
```

File: code/modules/lighting/lighting_turf.dm [hunk 2]
```diff
@@ -122,15 +112,28 @@
...
+		lighting_corner_SW = new/datum/lighting_corner(src, SOUTH|WEST)
+		if (recalc)
+			lighting_corner_SW.full_update()
+	
+	if (!lighting_corner_NW)
+		lighting_corner_NW = new/datum/lighting_corner(src, NORTH|WEST)
+		if (recalc)
+			lighting_corner_NW.full_update()
+	
+	lighting_corners_initialised = TRUE
```
Comment (MrStonedOne) [inline code/modules/lighting/lighting_turf.dm:139]:
```suggestion
	lighting_corners_initialised = TRUE

```

**Comment by MrStonedOne** [comment]
lighting init time is now 20% faster.

image

File: code/modules/lighting/lighting_corner.dm [hunk 3]
```diff
@@ -39,49 +40,55 @@ GLOBAL_LIST_INIT(LIGHTING_CORNER_DIAGONAL, list(NORTHEAST, SOUTHEAST, SOUTHWEST,
...
-	var/turf/T
+/datum/lighting_corner/proc/full_update()
 	var/thing
-	for (thing in masters)
-		T = thing
-		if (T.lighting_object)
-			active = TRUE
-			return
+	var/datum/light_source/L
+	for (thing in affecting)
```
Comment (Kylerace) [inline code/modules/lighting/lighting_corner.dm:89]: as anything

File: code/modules/lighting/lighting_source.dm [hunk 4]
```diff
@@ -231,50 +231,34 @@
...
-	affecting_turfs -= L
-	for (thing in L)
-		T = thing
-		LAZYREMOVE(T.affecting_lights, src)
-
+	var/list/L
 	LAZYINITLIST(effect_str)
 	if (needs_update == LIGHTING_VIS_UPDATE)
-		for (thing in  corners - effect_str) // New corners
+		for (thing in corners - effect_str) // New corners
```
Comment (Kylerace) [inline code/modules/lighting/lighting_source.dm:244]: all the loops that use thing can instead use as anything

File: code/modules/lighting/lighting_corner.dm [hunk 5]
```diff
@@ -134,6 +150,24 @@ GLOBAL_LIST_INIT(LIGHTING_CORNER_DIAGONAL, list(NORTHEAST, SOUTHEAST, SOUTHWEST,
 	if (!force)
 		return QDEL_HINT_LETMELIVE
 
-	stack_trace("Ok, Look, /tg/, I need you to find whatever fucker decided to call qdel on a fucking lighting corner, then tell him very nicely and politely that he is brain damaged and needs his head checked. Thanks. Send them my regards by the way.")
```
Comment (Cyberboss) [inline code/modules/lighting/lighting_corner.dm:137]: This is vintage PJB salt. Shame to lose it

**Comment by LemonInTheDark** [comment]
I'm currently tearing my hair out over a harddel caused by lighting datums being held inside lava's `affecting_lights` list. I'm just gonna pray this fixes it and move on

**Comment by optimumtact** [comment]
Why was this drafted it's been on basil for like weeks

**Comment by MrStonedOne** [comment]
> Why was this drafted it's been on basil for like weeks

breaks on new construction of lighting enabled tiles (ie, space -> floor tile)

**Comment by MrStonedOne** [comment]
> I'm currently tearing my hair out over a harddel caused by lighting datums being held inside lava's `affecting_lights` list. I'm just gonna pray this fixes it and move on

this pr removes that list.

File: code/modules/lighting/lighting_corner.dm [hunk 6]
```diff
@@ -39,49 +40,48 @@ GLOBAL_LIST_INIT(LIGHTING_CORNER_DIAGONAL, list(NORTHEAST, SOUTHEAST, SOUTHWEST,
...
 	// Now the horizontal one.
 	T = get_step(new_turf, horizontal)
 	if (T) // Ditto.
-		if (!T.corners)
-			T.corners = list(null, null, null, null)
-
-		masters[T]   = ((T.x > x) ? EAST : WEST) | ((T.y > y) ? NORTH : SOUTH) // Get the dir based on coordinates.
-		i            = GLOB.LIGHTING_CORNER_DIAGONAL.Find(turn(masters[T], 180))
-		T.corners[i] = src
+		save_master(T, ((T.x > x) ? EAST : WEST) | ((T.y > y) ? NORTH : SOUTH)) // Get the dir based on coordinates.
```
Comment (LemonInTheDark) [inline code/modules/lighting/lighting_corner.dm:52]: Would it be worth making this a define?

Comment (MrStonedOne) [reply code/modules/lighting/lighting_corner.dm:52]: My mind will melt if i stare at this code any longer, you can pr it to my branch or pr when you merge this.

Its a target for sure, now that corner datums get created later in the process and self clean up when lights go away.

Comment (LemonInTheDark) [reply code/modules/lighting/lighting_corner.dm:52]: aight, that's a good enough reason for me

**Comment by LemonInTheDark** [comment]
Lighting objects are hard deleting real bad, but not very often. I'm unsure of what could be causing this, but mso seemed to have some idea. <https://tgstation13.org/parsed-logs/terry/data/logs/2021/05/28/round-163093/runtime.log>
It's not common, but when it fires it's really bad. Wish I had the chance to sic the ref tracker on this

**Comment by YakumoChen** [comment]
Replacing tiles seems to fuck with the lighting in a way that triggers Nyctophobia even if you're in a lit area, I assume it's related.

**Comment by LemonInTheDark** [comment]
What server was this happening on?

**Pull request merged**